fsmonitor.c

```diff
@@ -173,27 +189,55 @@ void refresh_fsmonitor(struct index_state *istate)
 	 * should be inclusive to ensure we don't miss potential changes.
 	 */
 	last_update = getnanotime();
-	strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
+	if (hook_version == HOOK_INTERFACE_VERSION1)
+		strbuf_addf(&last_update_token, "%"PRIu64"", last_update);
 
 	/*
-	 * If we have a last update time, call query_fsmonitor for the set of
-	 * changes since that time, else assume everything is possibly dirty
+	 * If we have a last update token, call query_fsmonitor for the set of
+	 * changes since that token, else assume everything is possibly dirty
 	 * and check it all.
 	 */
 	if (istate->fsmonitor_last_update) {
-		query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION,
-			istate->fsmonitor_last_update, &query_result);
+		if (hook_version == -1 || hook_version == HOOK_INTERFACE_VERSION2) {
+			query_success = !query_fsmonitor(HOOK_INTERFACE_VERSION2,
+				istate->fsmonitor_last_update, &query_result);
+
+			if (query_success) {
+				/* First entry will be the last update token */
+				buf = query_result.buf;
+				bol = 0;
+				for (i = 0; i < query_result.len; i++) {
+					if (buf[i] == '\0') {
+						strbuf_addstr(&last_update_token, buf);
+						bol = i + 1;
+						break;
+					}
+				}	
```

Comment:
We know that `query_result` is NUL-terminated, so we can use all the amenities provided by `string.h`. I propose to write it this way instead:

```c
strbuf_addstr(&last_update_token, query_result.buf);
if (last_update_token.len == query_result.buf)
    [... handle invalid payload, probably by ignoring it somehow...]
bol = last_update_token.len + 1;
```

Reply:
Will you explain
`if (last_update_token.len == query_result.buf)`?
shouldn't it be 
`if (last_update_token.buf == query_result.buf)`

Reply:
Whoops. I think I meant to test for an empty token, i.e. `if (!last_update_token.len)`...